Don't issue a count for cached queries

AD7six opened the pull request:
I'm not sure about this but I noticed that a cached query, used in pagination, still issues a count query.

If the query was cached it's probably true that users wouldn't expect a query for the count to be issued either - so just count the results.

I can add some tests if I'm not missing something before merging.

AD7six commented:
I realize this is quite wrong.. since the count isn't meant to be the number of obtained rows but rather the number of matching rows ignoring limit/offset.

Is it worth allowing for caching the count of a query, or should users just define a counter callback to do it for them?

markstory commented on src/ORM/Query.php:
```diff
@@ -717,6 +717,10 @@ public function count()
             return (int)$counter($query);
         }
 
+        if (!empty($this->_cache)) {
```
Won't this cause issues when a query has caching on but isn't cached yet?

AD7six replied:
This isn't right, but no. The call to [QueryTrait::all](https://github.com/cakephp/cakephp/blob/ad-cache-count-too/src/Datasource/QueryTrait.php#L213-L215) would populate and return the cached value.

AD7six commented:
Taking a step back, right now it's not possible to avoid `$query->count()` issuing a query without registering a counter and implementing most of the code from `Query::count`, that'd look something like this:

    $query
        ->cache($cacheKey)
        ->counter(function($query) use ($countCacheKey) {
            return Cache::remember($countCacheKey, function()  use ($query) {
                $count = ['count' => $query->func()->count('*')];

                $statement = $query->connection()->newQuery()
                    ->select($count)
                    ->from(['count_source' => $query])
                    ->execute();

                $result = $statement->fetch('assoc')['count'];
                return (int)$result;
            }); 
        });

It's be nicer to be able to do:

    $query
        ->cache($cacheKey)
        ->counterCache($countCacheKey);

Or similar. worth implementing?

The count and the query cache keys should be disassociated since the count isn't dependent on e.g. the sort order/page wheras of course the results of the query are. The counterCache could default to `$cacheKey_count` though.

markstory commented:
If it was a separate method what wouxld the effect be if only counterCache was called?

AD7six commented:
In that case, only cache the count. Might be appropriate sometimes, e.g. large innodb tables; is just a suggestion though.

markstory commented:
Ok, would counterCache() also take cache configuration options? With it being independent it would have to do similar things to cache().

AD7six committed:
Only issue a count once

With code like this:

    $query->count();
    ...
    $query->count();

Only issue one count query

AD7six commented:
> Ok, would counterCache() also take cache configuration options? With it being independent it would have to do similar things to cache().

Sure.

But maybe it's not necessary, I realised my above example can actually be a whole lot simpler:

          $query
              ->cache($cacheKey);
              ->counter(function($query) use ($countCacheKey) {
                  return Cache::remember($countCacheKey, function()  use ($query) {
                      return $query->count();
                  }); 
              });

As the query class [wipes the counter object before running the callback](https://github.com/cakephp/cakephp/blob/master/src/ORM/Query.php#L716) which isn't quite trivial, but far from difficult to do. I'm not sure if it's worth implementing the counter cache I originally though was necessary, instead of just documenting how to use a counter callback.

Whilst poking around though I saw that:

    $query->count();
    $query->count();

Issues 2 queries - which is a simple thing to fix, and what I've repurposed this PR to add.

markstory commented on src/ORM/Query.php:
```diff
@@ -710,6 +719,10 @@ public function __clone()
     {
+        if (isset($this->_resultsCount)) {
+            return $this->_resultsCount;
+        }
```
I think this needs to check the dirty state. For example:

```php
$query->where(['id >' => 2]);
echo $query->count();
$query->where(['id <' => 1]);
echo $query->count();
```

Would now be wrong. I think this needs to be `if (isset($this->_resultsCount) && !$this->_dirty) {`

AD7six replied:
Nice catch, I'll add a test for that.

lorenzo commented:
@AD7six ping

AD7six committed:
Only if the query object is not dirty

AD7six committed:
Split the actual count into a seperate method

AD7six committed:
Add tests for count-cache

And account for the query being dirty too

AD7six committed:
When a query is executed - it's not dirty any more

AD7six committed:
Use the dirty function to clear the class property

AD7six committed:
Woot, that's now handled

It doesn't matter if the query is dirty - if it's not been modified
since the previous count it's still valid

AD7six commented:
Travis build problems excepted, this should be gtg

lorenzo commented:
Good job @AD7six
